File: backend/app/modules/schedules/domain/break_policy.py
"""Politique de pauses — normalisation multi-types (payées / non payées)."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Sequence

BreakKind = Literal["short", "meal", "other"]

INDUSTRIAL_2X10_MEAL_30: tuple[dict[str, Any], ...] = (
    {"minutes": 10, "paid": True, "kind": "short"},
    {"minutes": 10, "paid": True, "kind": "short"},
    {"minutes": 30, "paid": False, "kind": "meal"},
)


def normalize_break_item(raw: dict[str, Any]) -> dict[str, Any]:
    kind = str(raw.get("kind") or "other").strip().lower()
    if kind not in ("short", "meal", "other"):
        kind = "other"
    minutes = int(raw.get("minutes") or 0)
    return {
        "minutes": max(0, minutes),
        "paid": bool(raw.get("paid", False)),
        "kind": kind,
    }


def compute_break_totals(breaks: Sequence[dict[str, Any]]) -> tuple[int, int, int]:
    """Retourne (total, paid_minutes, unpaid_minutes)."""
    paid = 0
    unpaid = 0
    for item in breaks:
        norm = normalize_break_item(item)
        if norm["paid"]:
            paid += norm["minutes"]
        else:
            unpaid += norm["minutes"]
    return paid + unpaid, paid, unpaid


def breaks_from_legacy(break_minutes: int, break_paid: bool) -> list[dict[str, Any]]:
    if break_minutes <= 0:
        return []
    return [
        {
            "minutes": int(break_minutes),
            "paid": bool(break_paid),
            "kind": "meal" if not break_paid else "other",
        }
    ]


def resolve_breaks(raw: dict[str, Any]) -> list[dict[str, Any]]:
    """Résout breaks[] ou dérive depuis break_minutes / break_paid legacy."""
    raw_breaks = raw.get("breaks")
    if isinstance(raw_breaks, list) and raw_breaks:
        return [normalize_break_item(b) for b in raw_breaks if isinstance(b, dict)]
    legacy_min = int(raw.get("break_minutes") or 0)
    legacy_paid = bool(raw.get("break_paid", False))
    return breaks_from_legacy(legacy_min, legacy_paid)


def enrich_day_config_breaks(cfg: dict[str, Any]) -> dict[str, Any]:
    """Ajoute breaks[] et totaux dérivés à une config jour normalisée."""
    breaks = resolve_breaks(cfg)
    total, paid, unpaid = compute_break_totals(breaks)
    out = dict(cfg)
    out["breaks"] = breaks
    out["paid_break_minutes"] = paid
    out["unpaid_break_minutes"] = unpaid
    # Alias legacy : somme totale ; break_paid = toutes payées si total > 0
    out["break_minutes"] = total if total > 0 else int(cfg.get("break_minutes") or 0)
    if breaks:
        out["break_paid"] = paid > 0 and unpaid == 0
    return out


__all__ = [
    "INDUSTRIAL_2X10_MEAL_30",
    "BreakKind",
    "breaks_from_legacy",
    "compute_break_totals",
    "enrich_day_config_breaks",
    "normalize_break_item",
    "resolve_breaks",
]

File: frontend/src/lib/breakPolicy.ts
export type BreakKind = 'short' | 'meal' | 'other';

export interface DayBreak {
  minutes: number;
  paid: boolean;
  kind: BreakKind;
}

export const INDUSTRIAL_2X10_MEAL_30: DayBreak[] = [
  { minutes: 10, paid: true, kind: 'short' },
  { minutes: 10, paid: true, kind: 'short' },
  { minutes: 30, paid: false, kind: 'meal' },
];

export function computeBreakTotals(breaks: DayBreak[]): {
  total: number;
  paid: number;
  unpaid: number;
} {
  let paid = 0;
  let unpaid = 0;
  for (const b of breaks) {
    if (b.paid) paid += b.minutes;
    else unpaid += b.minutes;
  }
  return { total: paid + unpaid, paid, unpaid };
}

export function breaksFromLegacy(breakMinutes: number, breakPaid: boolean): DayBreak[] {
  if (breakMinutes <= 0) return [];
  return [
    {
      minutes: breakMinutes,
      paid: breakPaid,
      kind: breakPaid ? 'other' : 'meal',
    },
  ];
}

export function resolveBreaks(raw: {
  breaks?: DayBreak[];
  break_minutes?: number;
  break_paid?: boolean;
}): DayBreak[] {
  if (raw.breaks && raw.breaks.length > 0) {
    return raw.breaks.map((b) => ({
      minutes: Number(b.minutes) || 0,
      paid: Boolean(b.paid),
      kind: (b.kind as BreakKind) || 'other',
    }));
  }
  return breaksFromLegacy(Number(raw.break_minutes) || 0, Boolean(raw.break_paid));
}

export function syncLegacyBreakFields(breaks: DayBreak[]): {
  break_minutes: number;
  break_paid: boolean;
} {
  const { total, paid, unpaid } = computeBreakTotals(breaks);
  return {
    break_minutes: total,
    break_paid: total > 0 && unpaid === 0,
  };
}

export function grossPresenceHours(
  start: string,
  end: string,
  unpaidMinutes: number,
): number | null {
  if (!start || !end) return null;
  const [sh, sm] = start.split(':').map(Number);
  const [eh, em] = end.split(':').map(Number);
  if ([sh, sm, eh, em].some((n) => Number.isNaN(n))) return null;
  const gross = eh * 60 + em - (sh * 60 + sm);
  if (gross <= 0) return null;
  return Math.round(((gross - unpaidMinutes) / 60) * 100) / 100;
}

export function formatBreakSummary(breaks: DayBreak[]): string {
  const { paid, unpaid } = computeBreakTotals(breaks);
  const parts: string[] = [];
  if (paid > 0) parts.push(`${paid} min payées`);
  if (unpaid > 0) parts.push(`${unpaid} min repas`);
  return parts.length ? parts.join(' · ') : 'Aucune pause';
}
